langchain/vectorstores/vector_store_from_doc_index.py

```diff
@@ -0,0 +1,189 @@
+from operator import itemgetter
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Type
+
+try:
+    from docarray import BaseDoc
+    from docarray.index.abstract import BaseDocIndex
+    from docarray.typing import NdArray
+except ImportError:
+    BaseDoc = None
+    BaseDocIndex = None
+    NdArray = None
+
+from langchain.embeddings.base import Embeddings
+from langchain.schema import Document
+from langchain.vectorstores import VectorStore
+from langchain.vectorstores.utils import maximal_marginal_relevance
+
+
+def _check_docarray_import() -> None:
+    try:
+        import docarray
+
+        da_version = docarray.__version__.split(".")
+        if int(da_version[0]) == 0 and int(da_version[1]) <= 30:
+            raise ValueError(
+                f"To use the DocArrayHnswSearch VectorStore the docarray version >=0.31.0 is expected, "
+                f"received: {docarray.__version__}."
+                f"To upgrade, please run: `pip install -U docarray`."
+            )
+    except ImportError:
+        raise ImportError(
+            "Could not import docarray python package. "
+            "Please install it with `pip install \"langchain[docarray]\"`."
+        )
+
+
+class VecStoreFromDocIndex(VectorStore):
+    doc_index: BaseDocIndex
```

Comment:
VectorStore isn't currently a pedantic model, so these would actually be class variables not instance variables. but also since they're set in init, no need to have them here